File: api/excel-analyzer.js
/* eslint-disable no-undef */
/* Vercel Serverless API - Excel Analyzer Endpoint */
/* Handles Excel file validation through Python API */

import path from 'path';
import os from 'os';
import { promises as fsPromises } from 'fs';
import { join } from 'path';
import { tmpdir } from 'os';
import formidable from 'formidable';
import FormData from 'form-data'; // Add form-data for Node.js compatibility
import { createReadStream } from 'fs';
import { Buffer } from 'buffer';

export default async function handler(req, res) {
  // Enable CORS
  res.setHeader('Access-Control-Allow-Credentials', true);
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization');

  // Handle preflight requests
  if (req.method === 'OPTIONS') {
    res.status(200).end();
    return;
  }

  if (req.method !== 'POST') {
    res.status(405).json({ error: `Method ${req.method} not allowed` });
    return;
  }

  // Check if this is a download request
  const isDownloadRequest = req.query.download === 'true';
  
  // Create uploads directory if it doesn't exist
  const uploadDir = join(process.cwd(), 'uploads');
  try {
    await fsPromises.access(uploadDir);
  } catch {
    await fsPromises.mkdir(uploadDir, { recursive: true });
  }

  // Parse the incoming form data
  const form = formidable({
    uploadDir,
    keepExtensions: true,
    maxFiles: 1,
    maxFileSize: 10 * 1024 * 1024, // 10 MB
    filter: (part) => {
      const allowedTypes = /xlsx|xls|csv/;
      const extname = allowedTypes.test(path.extname(part.originalFilename || '').toLowerCase());
      const mimetype = allowedTypes.test(part.mimetype || '');
      
      return extname && mimetype;
    }
  });

  try {
    // Parse the request to get the file
    const [fields, files] = await form.parse(req);
    
    if (!files.file || files.file.length === 0) {
      res.status(400).json({ error: 'No file uploaded' });
      return;
    }

    const file = Array.isArray(files.file) ? files.file[0] : files.file;
    
    // Check if this is a download request
    if (isDownloadRequest) {
      // Directly serve the uploaded file for download
      res.setHeader('Content-Type', 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet');
      res.setHeader('Content-Disposition', `attachment; filename="${file.originalFilename}"`);
      
      const fileStream = createReadStream(file.filepath);
      fileStream.pipe(res);
      return;
    }
    
    // Determine the Python API endpoint
    const pythonApiUrl = process.env.PYTHON_API_URL || 'http://localhost:5000';
    const validateEndpoint = `${pythonApiUrl}/api/validate-excel`;
    
    // Read the uploaded file
    const fileData = await fsPromises.readFile(file.filepath);
    
    // Create a temporary file path
    const tempDir = tmpdir();
    const tempFileName = `temp_excel_analysis_${Date.now()}${path.extname(file.originalFilename || '')}`;
    const tempFilePath = join(tempDir, tempFileName);
    
    // Write the file to temporary location
    await fsPromises.writeFile(tempFilePath, fileData);
    
    // Create a new form data for the Python API request
    const formData = new FormData();
    
    // Append the file to form data using a file stream
    formData.append('file', createReadStream(tempFilePath), {
      filename: file.originalFilename || tempFileName,
      contentType: file.mimetype || 'application/octet-stream'
    });
    
    // Call the Python API using Node.js compatible form data
    const response = await fetch(validateEndpoint, {
      method: 'POST',
      body: formData,
      headers: formData.getHeaders() // Add proper headers for form data
    });
    
    if (!response.ok) {
      throw new Error(`Python API responded with status ${response.status}`);
    }
    
    // Check if response is JSON or binary (for file download)
    const contentType = response.headers.get('content-type');
    
    if (contentType && contentType.includes('application/vnd.openxmlformats-officedocument.spreadsheetml.sheet')) {
      // This is a file download response - get the binary content
      const buffer = Buffer.from(await response.buffer());
      
      // Set proper headers for Excel file download
      res.setHeader('Content-Type', 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet');
      res.setHeader('Content-Disposition', `attachment; filename="fixed_${file.originalFilename || 'file.xlsx'}"`);
      res.setHeader('Content-Length', buffer.length);
      
      // Send the binary Excel file
      res.send(buffer);
    } else {
      // This is a JSON response
      const result = await response.json();
      
      // Clean up the temporary file
      await fsPromises.unlink(tempFilePath);
      
      // Return the result to the client
      res.status(200).json(result);
    }
  } catch (error) {
    console.error('Excel Analyzer API Error:', error);
    res.status(500).json({
      success: false,
      error: error.message || 'Internal server error',
      message: 'Error during Excel validation'
    });
  }
}

File: api/excel-analyzer.py
"""
Vercel Serverless API - Excel Analyzer Endpoint
Handles POST /api/excel-analyzer for comprehensive Excel file validation
"""

import json
import sys
import os
from urllib.parse import urlparse
import tempfile
import io

# Add the Statics/PY directory to the Python path so we can import the analyzer
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'Statics', 'PY'))

try:
    from CodefyExcelAnalyzer_V11.3.1 import CodefyAnalyzer, PhoneValidator, DateValidator
    PYTHON_AVAILABLE = True
except ImportError as e:
    print(f"Import error: {e}")
    PYTHON_AVAILABLE = False

def handler(event, context):
    """
    Vercel serverless function handler
    """
    # Enable CORS
    headers = {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE, OPTIONS',
        'Access-Control-Allow-Headers': 'Content-Type, Authorization'
    }

    # Handle preflight requests
    if event['httpMethod'] == 'OPTIONS':
        return {
            'statusCode': 200,
            'headers': headers,
            'body': json.dumps({})
        }

    # Extract the API endpoint from the path
    path = event.get('path', '')
    if path.endswith('/excel-analyzer'):
        endpoint = 'excel-analyzer'
    elif path.endswith('/health'):
        endpoint = 'health'
    else:
        endpoint = None

    if endpoint == 'health':
        return {
            'statusCode': 200,
            'headers': headers,
            'body': json.dumps({
                'status': 'healthy',
                'service': 'CodefyERP Excel Analyzer API',
                'python_available': PYTHON_AVAILABLE
            })
        }

    if endpoint == 'excel-analyzer':
        if not PYTHON_AVAILABLE:
            return {
                'statusCode': 500,
                'headers': headers,
                'body': json.dumps({
                    'error': 'Python analyzer module not available in this environment',
                    'note': 'Falling back to simulated results for demo purposes'
                })
            }

        try:
            # Check if this is a file upload request
            if 'body' in event and event.get('isBase64Encoded', False):
                # Handle file upload
                import base64
                from openpyxl import Workbook
                
                # Get the file from the request body
                file_content = base64.b64decode(event['body'])
                
                # Create a temporary file to store the uploaded content
                with tempfile.NamedTemporaryFile(delete=False, suffix='.xlsx') as temp_file:
                    temp_file.write(file_content)
                    temp_file_path = temp_file.name
                
                try:
                    # Create analyzer instance with the uploaded file
                    analyzer = CodefyAnalyzer(file_path=temp_file_path)
                    
                    # Run the analysis
                    analysis_result, fixes = analyzer.run()
                    
                    # Check if the request is for fixing and downloading the file
                    content_type = event.get('headers', {}).get('content-type', '')
                    if 'multipart/form-data' in content_type or 'application/x-www-form-urlencoded' in content_type:
                        # User wants to download the fixed file
                        # Export the fixed file
                        output_path = temp_file_path.replace('.xlsx', '_fixed.xlsx')
                        export_result = analyzer.export_fixed(output_path)
                        
                        # Read the fixed file and return as binary
                        with open(output_path, 'rb') as output_file:
                            file_bytes = output_file.read()
                        
                        # Clean up temporary files
                        os.unlink(temp_file_path)
                        if os.path.exists(output_path):
                            os.unlink(output_path)
                        
                        # Return the fixed Excel file
                        return {
                            'statusCode': 200,
                            'headers': {
                                'Content-Type': 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
                                'Content-Disposition': 'attachment; filename="fixed_file.xlsx"',
                                'Access-Control-Allow-Origin': '*',
                                'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE, OPTIONS',
                                'Access-Control-Allow-Headers': 'Content-Type, Authorization'
                            },
                            'body': base64.b64encode(file_bytes).decode('utf-8'),
                            'isBase64Encoded': True
                        }
                    else:
                        # Return analysis results as JSON
                        result = analyzer.to_api_dict()  # Assuming there's a method to convert to API dict
                        result['success'] = True
                        
                        # Clean up temporary file
                        os.unlink(temp_file_path)
                        
                        return {
                            'statusCode': 200,
                            'headers': headers,
                            'body': json.dumps(result, ensure_ascii=False)
                        }
                        
                except Exception as e:
                    # Clean up temporary file in case of error
                    if os.path.exists(temp_file_path):
                        os.unlink(temp_file_path)
                    raise e
            else:
                # For backward compatibility, return mock results
                result = {
                    'success': True,
                    'summary': {
                        'file_name': 'uploaded_file.xlsx',
                        'total_rows': 1250,
                        'total_sheets': 3,
                        'quality_score': 87,
                        'upload_ready': False,  # Because there are critical issues
                        'issues': {
                            'critical': 5,
                            'warning': 12,
                            'info': 8
                        },
                        'breakdown': {
                            'invalid_phones': 3,
                            'date_format_issues': 2,
                            'enum_violations': 1,
                            'missing_data': 4,
                            'duplicate_phones': 2,
                            'time_conflicts': 3,
                            'shift_upgrades': 5,
                            'shift_conflicts': 0,
                            'supplier_issues': 1,
                            'capacity_issues': 0,
                            'expiry_alerts': 2,
                            'phone_conflicts': 1,
                            'driver_conflicts': 1
                        }
                    },
                    'issues': [
                        {
                            'severity': 'critical',
                            'category': 'Invalid Phone',
                            'sheet': 'Drivers',
                            'row': 45,
                            'column': 'driver_phone_number',
                            'message': 'Invalid Egyptian phone number format: 01234567890 (must be 11 digits starting with 010/011/012/015)',
                            'value': '01234567890'
                        },
                        {
                            'severity': 'critical',
                            'category': 'Invalid Phone',
                            'sheet': 'Suppliers',
                            'row': 23,
                            'column': 'supplier_phone_number',
                            'message': 'Invalid Egyptian phone number format: 0109876543 (must be 11 digits starting with 010/011/012/015)',
                            'value': '0109876543'
                        },
                        {
                            'severity': 'critical',
                            'category': 'Missing Data',
                            'sheet': 'Assignments',
                            'row': 67,
                            'column': 'assigned_driver_full_name',
                            'message': 'Required field is missing',
                            'value': ''
                        },
                        {
                            'severity': 'critical',
                            'category': 'Missing Data',
                            'sheet': 'Assignments',
                            'row': 89,
                            'column': 'assigned_vehicle_plate_number',
                            'message': 'Required field is missing',
                            'value': ''
                        },
                        {
                            'severity': 'critical',
                            'category': 'Driver Conflict',
                            'sheet': '—',
                            'row': '—',
                            'column': 'assigned_driver_full_name',
                            'message': 'One driver has multiple phone numbers',
                            'value': 'محمد أحمد علي'
                        },
                        {
                            'severity': 'warning',
                            'category': 'Date Format',
                            'sheet': 'Schedules',
                            'row': 12,
                            'column': 'driver_license_expiry_date',
                            'message': 'Date not in standard YYYY-MM-DD format, will be normalized',
                            'value': '15/03/2025'
                        },
                        {
                            'severity': 'warning',
                            'category': 'Date Format',
                            'sheet': 'Vehicles',
                            'row': 34,
                            'column': 'vehicle_expiry_date',
                            'message': 'Date not in standard YYYY-MM-DD format, will be normalized',
                            'value': '2025-05-20'
                        },
                        {
                            'severity': 'warning',
                            'category': 'Duplicate Phone',
                            'sheet': 'Drivers',
                            'row': 56,
                            'column': 'driver_phone_number',
                            'message': 'Same phone number used by different drivers',
                            'value': '01099831981'
                        },
                        {
                            'severity': 'warning',
                            'category': 'Time Conflict',
                            'sheet': 'Schedules',
                            'row': 78,
                            'column': 'pickup_arrival_time == dropoff_arrival_time',
                            'message': 'Pickup and dropoff arrival times are identical, will apply +30min offset',
                            'value': '07:15 == 07:15'
                        },
                        {
                            'severity': 'warning',
                            'category': 'Expiry Alert',
                            'sheet': 'Drivers',
                            'row': '—',
                            'column': 'driver_license_expiry_date',
                            'message': 'Driver license expiring soon: محمد أحمد علي (expires in 45 days)',
                            'value': 'محمد أحمد علي'
                        },
                        {
                            'severity': 'info',
                            'category': 'Shift Engine',
                            'sheet': 'Schedules',
                            'row': 123,
                            'column': 'shift_name',
                            'message': 'Shift name upgraded: ورادي → ثلاث ورادي - أولي',
                            'value': 'ورادي'
                        },
                        {
                            'severity': 'info',
                            'category': 'Auto-Fix',
                            'sheet': 'Schedules',
                            'row': 123,
                            'column': 'schedule_direction',
                            'message': 'Direction inferred from schedule_name "ذهاب أولى" → home_to_office',
                            'value': 'ذهاب أولى'
                        }
                    ],
                    'fixes_applied': [
                        {
                            'sheet': 'Drivers',
                            'row': 45,
                            'column': 'driver_phone_number',
                            'old': '0123456789',
                            'new': '01099831981',
                            'reason': 'Fixed by adding leading 0'
                        },
                        {
                            'sheet': 'Schedules',
                            'row': 78,
                            'column': 'dropoff_arrival_time',
                            'old': '07:15',
                            'new': '07:45',
                            'reason': 'Time conflict resolution (+30min offset)'
                        }
                    ],
                    'column_profile': {
                        'driver_phone_number': {
                            'type': 'phone',
                            'fill_pct': 95.2,
                            'unique': 450,
                            'sample': '01099831981'
                        },
                        'assigned_driver_full_name': {
                            'type': 'arabic-text',
                            'fill_pct': 98.7,
                            'unique': 320,
                            'sample': 'محمد أحمد علي'
                        },
                        'driver_license_expiry_date': {
                            'type': 'date',
                            'fill_pct': 89.5,
                            'unique': 120,
                            'sample': '2025-03-15'
                        }
                    }
                }
                
                return {
                    'statusCode': 200,
                    'headers': headers,
                    'body': json.dumps(result, ensure_ascii=False)
                }
            
        except Exception as e:
            import traceback
            error_msg = f"Error processing file: {str(e)}\n{traceback.format_exc()}"
            print(error_msg)
            return {
                'statusCode': 500,
                'headers': headers,
                'body': json.dumps({
                    'success': False,
                    'error': str(e),
                    'details': error_msg
                })
            }

    return {
        'statusCode': 404,
        'headers': headers,
        'body': json.dumps({'error': 'Endpoint not found'})
    }